File: acp/acp-core/src/main/java/pers/acp/core/log/LogFactory.java
package pers.acp.core.log;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;

/**
 * Create by zhangbin on 2017-11-8 15:47
 */
public class LogFactory {

    private final Logger logger;

    private LogFactory(Class<?> clas) {
        logger = LoggerFactory.getLogger(clas);
    }

    private LogFactory(String name) {
        logger = LoggerFactory.getLogger(name);
    }

    public static LogFactory getInstance(Class<?> clas) {
        return new LogFactory(clas);
    }

    public static LogFactory getInstance(String name) {
        return new LogFactory(name);
    }

    public void info(String message) {
        setCustomerParams();
        logger.info(message);
    }

    public void info(String message, Object... var) {
        setCustomerParams();
        logger.info(message, var);
    }

    public void info(String message, Throwable t) {
        setCustomerParams();
        logger.info(message, t);
    }

    public void debug(String message) {
        setCustomerParams();
        logger.debug(message);
    }

    public void debug(String message, Object... var) {
        setCustomerParams();
        logger.debug(message, var);
    }

    public void debug(String message, Throwable t) {
        setCustomerParams();
        logger.debug(message, t);
    }

    public void warn(String message) {
        setCustomerParams();
        logger.warn(message);
    }

    public void warn(String message, Object... var) {
        setCustomerParams();
        logger.warn(message, var);
    }

    public void warn(String message, Throwable t) {
        setCustomerParams();
        logger.warn(message, t);
    }

    public void error(String message) {
        setCustomerParams();
        logger.error(message);
    }

    public void error(String message, Object... var) {
        setCustomerParams();
        logger.debug(message, var);
    }

    public void error(String message, Throwable t) {
        setCustomerParams();
        logger.error(message, t);
    }

    public void trace(String message) {
        setCustomerParams();
        logger.trace(message);
    }

    public void trace(String message, Object... var) {
        setCustomerParams();
        logger.trace(message, var);
    }

    public void trace(String message, Throwable t) {
        setCustomerParams();
        logger.trace(message, t);
    }

    private void setCustomerParams() {
        StackTraceElement[] stacks = Thread.currentThread().getStackTrace();
        int lineno = 0;
        if (stacks.length >= 4) {
            lineno = stacks[3].getLineNumber();
        }
        MDC.put("lineno", lineno + "");
    }

}

File: core/acp-core/src/main/kotlin/pers/acp/core/log/LogFactory.kt
package pers.acp.core.log

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.slf4j.MDC

/**
 * @author zhang by 10/07/2019
 * @since JDK 11
 */
class LogFactory {

    private var logger: Logger

    private var isCls: Boolean

    val stackIndex: Int

    private constructor(cls: Class<*>, stackIndex: Int = 3) {
        logger = LoggerFactory.getLogger(cls)
        this.stackIndex = stackIndex
        isCls = true
    }

    private constructor(name: String, stackIndex: Int = 3) {
        logger = LoggerFactory.getLogger(name)
        this.stackIndex = stackIndex
        isCls = false
    }

    fun info(message: String?) {
        setCustomerParams()
        logger.info(message)
    }

    fun info(message: String?, vararg variable: Any?) {
        setCustomerParams()
        logger.info(message, *variable)
    }

    fun info(message: String?, t: Throwable?) {
        setCustomerParams()
        logger.info(message, t)
    }

    fun debug(message: String?) {
        setCustomerParams()
        logger.debug(message)
    }

    fun debug(message: String?, vararg variable: Any?) {
        setCustomerParams()
        logger.debug(message, *variable)
    }

    fun debug(message: String?, t: Throwable?) {
        setCustomerParams()
        logger.debug(message, t)
    }

    fun warn(message: String?) {
        setCustomerParams()
        logger.warn(message)
    }

    fun warn(message: String?, vararg variable: Any?) {
        setCustomerParams()
        logger.warn(message, *variable)
    }

    fun warn(message: String?, t: Throwable?) {
        setCustomerParams()
        logger.warn(message, t)
    }

    fun error(message: String?) {
        setCustomerParams()
        logger.error(message)
    }

    fun error(message: String?, vararg variable: Any?) {
        setCustomerParams()
        logger.debug(message, *variable)
    }

    fun error(message: String?, t: Throwable?) {
        setCustomerParams()
        logger.error(message, t)
    }

    fun trace(message: String?) {
        setCustomerParams()
        logger.trace(message)
    }

    fun trace(message: String?, vararg variable: Any?) {
        setCustomerParams()
        logger.trace(message, *variable)
    }

    fun trace(message: String?, t: Throwable?) {
        setCustomerParams()
        logger.trace(message, t)
    }

    private fun setCustomerParams() {
        val stacks = Thread.currentThread().stackTrace
        var lineno = 0
        var className = ""
        if (stacks.size >= stackIndex + 1) {
            lineno = stacks[stackIndex].lineNumber
            className = stacks[stackIndex].className
        }
        if (isCls) {
            logger = LoggerFactory.getLogger(className)
        }
        MDC.put("lineno", lineno.toString())
    }

    companion object {

        @JvmStatic
        @JvmOverloads
        fun getInstance(cls: Class<*>, stackIndex: Int = 3): LogFactory {
            return LogFactory(cls, stackIndex)
        }

        @JvmStatic
        @JvmOverloads
        fun getInstance(name: String, stackIndex: Int = 3): LogFactory {
            return LogFactory(name, stackIndex)
        }
    }

}
